src/facet_display/FilterableFacet.tsx

```diff
@@ -0,0 +1,124 @@
+import React, { useState, useEffect, useCallback } from "react"
+import { contains } from "ramda"
+import Fuse from "fuse.js"
+
+import { SearchFacetItem } from "./SearchFacetItem"
+import { BucketWithLabel } from "./FacetDisplay"
+
+// the `.search method returns records like { item, refindex }
+// where item is the facet and refIndex is it's index in the original
+// array. this helper just pulls out only the facets themselves
+const runSearch = (searcher: Fuse<BucketWithLabel>, text: string) =>
+  searcher.search(text).map(({ item }) => item)
+
+interface Props {
+  name: string
+  title: string
+  results: BucketWithLabel[] | null
+  currentlySelected: string[]
+  onUpdate: React.ChangeEventHandler<HTMLInputElement>
+  expandedOnLoad: boolean
+}
+
+function FilterableFacet(props: Props) {
+  const { name, title, results, currentlySelected, onUpdate, expandedOnLoad } =
+    props
+  const [showFacetList, setShowFacetList] = useState(expandedOnLoad)
+
+  // null is signal for no input yet or cleared input
+  const [filteredList, setFilteredList] = useState<BucketWithLabel[] | null>(
+    null
+  )
+  const [searcher, setSearcher] = useState<Fuse<BucketWithLabel>>(new Fuse([]))
+
+  const [filterText, setFilterText] = useState("")
+
+  useEffect(() => {
```

Comment:
**React Suggestions:** If I look at the UI for this component, the data I see are:
1. **results**: (aka buckets) passed via parent.
2. **filterText:** entered by user, a state variable
3. **showFacetList**: controlled by user, a state variable (Suggest renaming to "expanded").

It seems to me that `filteredList` and `searcher` are not independent state, they are derived from other values. So let's not use `useState` for them. Instead:

```tsx
const [filterText, setFilterText] = useState("")
const filteredResults = useMemo(() => {
  return filterText ? runSearch(
    new Fuse(results, { keys: ["key", "label"], threshold: 0.4 }),
    filterText
  ) : results
)
```